### DISCUSSION TITLE
[amnb] build error with code from g071 example

### DISCUSSION THREAD

**Discussion by Tecnologic** [modm-io/modm#1215] [Q&A] [answered]
Hi all,

i'm quite new to modm an trying to evaluate if the amnb protocol is something for my application. With the great docs i started a project on the F4-disco board and used the nucleo-g071 amnb example as a base.

But when compiling i get this error: 
```
`main.cpp: In member function 'bool Thread::update()':
modm\src/modm/processing/protothread/macros.hpp:91:22: error: jump to case label
   91 |                 case __LINE__: \
      |                      ^~~~~~~~
main.cpp:111:32: note: in expansion of macro 'PT_CALL'
  111 |                         res1 = PT_CALL(node2.request<uint8_t>(1, 1));`
```

The code is a direct copy of the example:

```cpp
class Thread : public modm::pt::Protothread
{
	modm::ShortPeriodicTimer tmr{1s};
	uint32_t counter{0};
	Result<uint8_t> res1;
	Result<uint8_t, uint8_t> res2;

public:
	bool inline
	update()
	{
		PT_BEGIN();

		while(true)
		{
			PT_WAIT_UNTIL(tmr.execute());

			LedGreen::toggle();

			node1.broadcast(1, counter++);
			node3.broadcast(2);

			res1 = PT_CALL(node2.request<uint8_t>(1, 1));
			MODM_LOG_INFO << "Node1 responded with: " << res1.error();
			if (res1) { MODM_LOG_INFO << " " << *res1 << modm::endl; }

			res2 = PT_CALL(node1.request<uint8_t, uint8_t>(3, 2, counter));
			MODM_LOG_INFO << "Node3 responded with: " << res2.error();
			if (res2.hasUserError()) {
				MODM_LOG_INFO << " " << *res2.userError() << modm::endl;
			}
		}

		PT_END();
	}
}
thread;
```

I cant see the error here. PT_BEGIN is where it should be, i see no declaration of a local variable which could be uninitialized in a next case, in short i dot get what gcc is saying to me. 

Is this a potential bug or is it me doing something wrong with protothreads? 

Thanks

**Comment by salkinium** [comment]
Are you using GCC13 by any chance? Our protothreads macros only work with ≤GCC12 sadly.
We're therefore moving away to [fibers](https://modm.io/reference/module/modm-processing-fiber/) for which there exists a [backward compatible option for protothreads](https://modm.io/reference/module/modm-processing-protothread/#using-fibers) and [for resumable functions](https://modm.io/reference/module/modm-processing-resumable/#using-fibers).

Try to either use GCC12 or enable the `modm:processing:protothread:use_fiber` option.

**Reply by Tecnologic** [reply]
Hi Thanks for the answer.

yes i use gcc13. 

I added `<option name="modm:processing:protothread:use_fiber">True</option>` and `<module>modm:processing:resumable</module>`, cleaned lbuild and scons but i keep getting 
```
`C:/Program Files (x86)/Arm GNU Toolchain arm-none-eabi/13.3 rel1/bin/../lib/gcc/arm-none-eabi/13.3.1/../../../../arm-none-eabi/bin/ld.exe: build\test\scons-debug\main.o: in function `modm::amnb::Node<2u, 0u>::transmit()':
./modm\src/modm/communication/amnb/node.hpp:144:(.text._ZN4modm4amnb4NodeILj2ELj0EE8transmitEv[_ZN4modm4amnb4NodeILj2ELj0EE8transmitEv]+0xe): undefined reference to `modm::Resumable<(unsigned char)6>::isResumableRunning(unsigned char) const'
C:/Program Files (x86)/Arm GNU Toolchain arm-none-eabi/13.3 rel1/bin/../lib/gcc/arm-none-eabi/13.3.1/../../../../arm-none-eabi/bin/ld.exe: build\test\scons-debug\main.o: in function `modm::amnb::Node<2u, 0u>::request()':
./modm\src/modm/communication/amnb/node.hpp:187:(.text._ZN4modm4amnb4NodeILj2ELj0EE7requestEv[_ZN4modm4amnb4NodeILj2ELj0EE7requestEv]+0xa): undefined reference to `modm::Resumable<(unsigned char)6>::isResumableRunning(unsigned char) const'`
```

how can i get lbuild to implement `modm::Resumable`?

**Reply by Tecnologic** [reply]
`// unimplementable with fibers, but may be stubbed by user application
    void stopAllResumables();
    bool stopResumable(uint8_t id);
    bool isResumableRunning(uint8_t id) const;`

reading that if added:

`template< uint8_t Functions >
bool modm::Resumable<Functions>::isResumableRunning(uint8_t id) const
{
	(void)id;
	return false;
}`

to my main.cpp and it builds

**Reply by salkinium** [reply]
Ooops, that's right, I completely forgot that AMNB is using `isResumableRunning()`…
I've [opened a PR that replaces the usage with the correct implementation](https://github.com/modm-io/modm/pull/1216) that should be compatible with fibers.
With this fix you should be able to request/broadcast from different fibers without mutual exclusion issues.

**Reply by Tecnologic** [reply]
Thanks,

with the current code from the example for the g071 there is no call to the fiber scheduler. I think this isnt. right? 
So the example should be ported from the protothreads implementation to fibers? 

i moved the update calles from the main while loop to fibers:
```cpp
modm::Fiber<> fiberNode1([]{node1.update(); modm::this_fiber::yield();});
modm::Fiber<> fiberNode2([]{node2.update(); modm::this_fiber::yield();});
modm::Fiber<> fiberNode3([]{node3.update(); modm::this_fiber::yield();});
```

and added the scheduler call to the end of main.

now i'm struck at the second request call 
`RF_WAIT_WHILE(interface.isMediumBusy());`
It seems like the other fibers are not executed but this RF_WAIT_WHILE should yield.

sry for all my questions, but i find it a good problem to get familiar with modm :)

**Reply by salkinium** [reply]
Ah sorry, no. The `use_fiber` option automatically schedules the protothreads on construction, so you should only have to replace the `while(true)` loop with the `modm::fiber::Scheduler::run();`.  But there was another bug in the where the AMNB `update()` function got stuck in a (unnecessary) while(1) loop. I've updated [my PR](https://github.com/modm-io/modm/pull/1216) with another change that should unblock the fibers.

> sry for all my questions, but i find it a good problem to get familiar with modm :)

No worries, I'm also interested in getting this to work with fibers! I initially wrote the AMNB for [this project](https://chaos.social/@salkinium/113313183794636898) and achieved very throughput good results even in congestion.
But I think the code could be simplified with a proper port from protothreads to fibers. I'll hack on it more properly this weekend.

**Reply by salkinium** [reply]
Btw I tried to get this to work with fibers and it's indeed getting stuck somewhere. I think there isn't an easy fix, I'll have to debug this further. I would suggest you downgrade to GCC12 and use Protothreads for your evaluation.

**Reply by Tecnologic** [reply]
Thanks, I will continue first with pwm, ADC and then revisit this later

**Reply by salkinium** [reply, marked as answer]
I fixed it in [the PR](https://github.com/modm-io/modm/pull/1216), turns out I implemented AMNB with two threads one for transmission and one for reception.

**Reply by Tecnologic** [reply]
I try to pull the pr and test it out!

**Comment by Tecnologic** [comment]
I read from that, that fiber is the way to go for my application?

**Reply by salkinium** [reply]
It depends on how much RAM you have available. Stackful fibers are more ergonomic since you can use local variables, but they require a lot more stack memory (particularly if you use printf), thus might not be great on smaller devices.
I recommend you measure the stack usage with [the watermarking feature](https://modm.io/reference/module/modm-processing-fiber/#stack-usage).

**Reply by Tecnologic** [reply]
Thanks for the reply.

I want to port my foc esc project from chibios rtos to modm, because i dont need a rtos for all the stuff but some sheduling to keep the code neat and tidy is nice. So fibers seem a good alternative for all the non control stuff.
